pgrx/src/list.rs

```diff
@@ -12,257 +12,617 @@
+        /// Due to lifetimes this isn't a problem until unsafe Rust becomes involved,
+        /// but with Postgres extensions it often does.
+        pub fn as_cells(&self) -> &[ListCell<T>] {
+            unsafe { slice::from_raw_parts(self.as_cells_ptr(), self.len()) }
+        }
+
+        pub fn as_cells_ptr(&self) -> *const ListCell<T> {
+            unsafe { (*self.list.as_ptr()).elements.cast() }
+        }
+
+        pub fn as_mut_cells_ptr(&mut self) -> *mut ListCell<T> {
+            unsafe { (*self.list.as_ptr()).elements.cast() }
         }
     }
 
-    #[cfg(any(feature = "pg13", feature = "pg14", feature = "pg15", feature = "pg16"))]
-    #[inline]
-    pub fn replace_oid(&mut self, i: usize, with: pg_sys::Oid) {
-        unsafe {
-            let cell = pg_sys::pgrx_list_nth_cell(self.list, i as i32);
-            cell.as_mut().expect("cell is null").oid_value = with;
+    impl<T: Enlist> ListHead<T> {
+        /// From a non-nullable pointer that points to a valid List, produce a ListHead of the correct type
+        ///
+        /// # Safety
+        /// This assumes the NodeTag is valid to read, so it is not okay to call this on
+        /// pointers to deallocated or uninit data.
+        ///
+        /// If it returns as `Some`, it also asserts the entire List is, across its length,
+        /// validly initialized as `T` in each ListCell.
+        pub unsafe fn downcast_from_ptr(list: NonNull<pg_sys::List>) -> Option<ListHead<T>> {
+            (T::LIST_TAG == (*list.as_ptr()).type_).then_some(ListHead { list, _type: PhantomData })
+        }
+
+        pub fn push(&mut self, value: T) -> &mut Self {
+            let list = unsafe { self.list.as_mut() };
+            let pg_sys::List { length, max_length, elements, .. } = list;
+            if *max_length - *length > 0 {
+                // SAFETY: Our list must have been constructed following the list invariants
+                // in order to actually get here, and we have confirmed as in-range of the buffer.
+                let cell = unsafe { &mut *elements.add(*length as _) };
+                T::endocytosis(cell, value);
+                *length += 1;
+            } else {
+                // Reserve in this branch.
+                let new_cap = max_length.saturating_mul(2);
+                self.reserve(new_cap as _);
+            }
+
+            // Return `self` for convenience of `List::try_push`
+            self
+        }
+
+        pub fn reserve(&mut self, size: usize) -> &mut Self {
+            let list = unsafe { self.list.as_mut() };
+            if ((list.max_length - list.length) as usize) < size {
+                unsafe { grow_list(list, size + list.length as usize) };
+            };
+            self
         }
     }
 
-    #[inline]
-    pub fn iter_ptr(&self) -> impl Iterator<Item = *mut T> + '_ {
-        PgListIteratorPtr { list: &self, pos: 0 }
+    unsafe fn grow_list(list: &mut pg_sys::List, target: usize) {
```

Comment:
This doesn't cover the "debug" branches that the Postgres version has (DEBUG_LIST_MEMORY_USAGE) but I think that's okay.